pkg/ddl/partition.go

```diff
@@ -3801,59 +3852,91 @@ func (w *reorgPartitionWorker) BackfillData(handleRange reorgBackfillTask) (task
 		}
 		txn.SetOption(kv.ResourceGroupName, w.jobContext.resourceGroupName)
 
-		rowRecords, nextKey, taskDone, err := w.fetchRowColVals(txn, handleRange)
+		nextKey, taskDone, err := w.fetchRowColVals(txn, handleRange)
 		if err != nil {
 			return errors.Trace(err)
 		}
 		taskCtx.nextKey = nextKey
 		taskCtx.done = taskDone
 
-		warningsMap := make(map[errors.ErrorID]*terror.Error)
-		warningsCountMap := make(map[errors.ErrorID]int64)
-		for _, prr := range rowRecords {
-			taskCtx.scanCount++
+		isClustered := w.reorgedTbl.Meta().IsCommonHandle || w.reorgedTbl.Meta().PKIsHandle
+		if !isClustered {
+			// non-clustered table, we need to replace the _tidb_rowid handles since
+			// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
+			// Meaning we need to check here if a record was double written to the new partition,
+			// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
+			// and we should skip it.
+			var found map[string][]byte
+			if len(w.oldKeys) > 0 {
+				// we must check if old IDs already been written,
+				// i.e. double written by StateWriteOnly or StateWriteReorganization.
```

Comment:
It means that here we need to check in the new partitions, if the read rows exists there already (with the same _tidb_rowid). I.e. we first read from the old partition, calculate the new partition id, and from all the read rows from the old partitions, does a BatchGet from those rows but in the new partitions. If any of those rows exists, we will skip them, since those were double written during insert/update and can therefore be skipped.

So short answer, we did not modify those rows in the DDL, but a concurrent user session did it before we came to backfill from the old partitions to the new.